feature: verify shelter

eTaphee opened the pull request:
<!-- 필요한 경우 팀원들과 의논하여 양식을 변경할 수 있습니다. -->
<!-- PR은 상세히 적어주시는게 좋습니다. -->
<!-- 이해를 돕기위한 이미지를 써도 좋습니다. -->

## Changes
<!-- 어떤점이 변경되었는지 적어주세요. -->
보호소 승인 기능 구현
- ADMIN 권한이 있는 사용자만 접근할 수 있어야 합니다.
- 승인된 보호소를 다시 호출하면 예외 발생
- 승인 시 verified 필드 업데이트 및 SHELTER 권한을 추가합니다.

## Background
<!-- 이 PR이 진행된 배경을 적어주세요. -->
보호소 등록 후 승인된 보호소만 분양게시글을 작성할 수 있도록 보호소 승인 API가 필요합니다.

## Discuss
<!-- 토론할 내용이 있다면 적어주세요. -->
#58  브랜치를 베이스로 작업을 했지만 아직 승인이 되지 않아 develop 브랜치에 머지가 안 된 상태입니다.
그래서 develop 대상으로 pr요청하면 #58 브랜치에서 수정한 내용이 같이 포함되기 때문에 베이스 브랜치로 머지하는 PR 작성합니다.

SHELTER에서 바로 사용자에 접근할 수 있도록 양방향 매핑으로 변경했습니다. ERD도 수정완료했습니다.
혹시 문제가 될 부분이 있다고 생각드시면 말씀해주세요.

## Issues
<!-- 관련된 이슈를 #{issueNumber}를 통해 태그해주세요. -->
<!-- ex) 해결한 이슈: #1, #2 -->
<!-- ex) 구현이 필요한 이슈: #3, #4 -->
해결한 이슈: #58

## Execute
<!-- 실행된 결과에 대해 적어주세요. -->
<!-- 어떻게 테스트를 진행했는지, 어떤 검토를 거쳤는지 적어주시면 좋습니다. -->
<!-- 테스트 코드는 중요합니다! -->

## Reference
<!-- 참조한 레퍼런스가 있다면 적어주세요. -->
<img width="1125" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/d0bd8de5-fb88-4299-a5f7-a8dc589b3698">

<img width="1125" alt="image" src="https://github.com/banzzokee/banzzokee-back/assets/43163409/72c7be62-8e8d-4b0c-8403-ff1bb5b4074c">

eTaphee committed:
feature: 보호소 승인 구현

eTaphee committed:
test: 테스트 컨테이너 비활성

eTaphee committed:
test: 테스트 코드 수정

Shelter에 UserId 관계를 추가하여 영향받는 테스트 코드 수정

eTaphee committed:
test: 보호소 승인 테스트 작성

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/exception/ShelterAlreadyVerifiedException.java:
```diff
@@ -0,0 +1,17 @@
+@Getter
+public class ShelterAlreadyVerifiedException extends CustomException {
+
+  private final Long shelterId;
```
shelterId가 따로 쓰이는 곳이 있는건가요???

eTaphee replied:
이전 피드 참고해서 지우도록 하겠습니다~

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/entity/Shelter.java:
```diff
@@ -66,10 +74,14 @@ public class Shelter extends BaseEntity {
 
+  @OneToOne(fetch = LAZY)
+  @JoinColumn(name = "user_id")
+  private final User user;
```
양방향으로 수정하셨군요..!

혹시 제가 말씀드린거 때문에 수정하신거면
이방법이 찝찝하진 않은지 다른 분들이랑 확실히 얘기 나눠보고 다시 단방향으로 진행해도 괜찮습니다.
말씀드릴때 의견종합이 확실히 안되었던걸로 기억합니다...

eTaphee replied:
이 부분은 논의를 해봐야겠네요~

jnjeaaaat commented on src/main/java/homes/banzzokee/domain/shelter/service/ShelterService.java:
```diff
@@ -38,20 +46,50 @@ public void registerShelter(ShelterRegisterRequest request, MultipartFile shelte
+  }
+
   private void throwIfUserAlreadyRegisterShelter(User user) {
     if (user.hasShelter()) {
```
User가 Shelter를 이미 등록했는지 여부를
SHELTER 권한 유무로 체크하셨는데 아래와 같은 문제가 발생할 듯 합니다.

1. 보호소 등록 신청 -> 이미 신청했는지 확인(role에 SHELTER 없음 - 통과)
-> Shelter db에 save(아직 없는 코드?) -> User entity에 승인안난 Shelter 등록

2-1. 다시 보호소 등록 신청 -> 이미 신청했는지 확인(role에 SHELTER 없음 - 통과)
-> 똑같은 Shelter db에 save

2-2. hasShelter()가 SHELTER 권한 유무가 아닌
if (user.getShelter != null && !user.getShelter.verified())
-> throw new 승인필요함()

eTaphee replied:
그렇게 생각할 수도 있네요.
제가 role 검사로만 판단했던 이유는..

shelter를 등록했는지 여부를 확인하기 위해 한번더 DB 쿼리를 하는게 싫어서(?)였습니다.
그래서 verify이 후 SHELTER role을 추가했습니다.

태준님 리뷰를 보고 다시 코드를 보니 role로만 검사하게 되는 경우 등록 시 문제가 생기겠네요.

놓친부분 짚어주셔서 감사합니다.

eTaphee committed:
Merge branch 'feature/register-shelter' into feature/approve-shelter

eTaphee committed:
feature: 보호소 수정 구현

eTaphee committed:
test: 보호소 수정 테스트 작성

eTaphee committed:
feature: 보호소 삭제 구현

- 삭제된 보호소에 대해 인증이 되지 않도록 수정
- 보호소를 삭제 후 다시 등록될 수 있도록 구현
- 보호소 삭제 로직
  - 보호소의 deletedAt, verified 업데이트
  - 사용자의 SHELTER 권한 삭제
  - 보호소가 참여한 채팅방에서 shelter_id = null 업데이트

eTaphee committed:
test: 보호소 삭제 테스트 작성

eTaphee committed:
refactor: 보호소 서비스 테스트 리팩터링

eTaphee committed:
test: 보호소 승인 테스트 케이스 추가

nadoran78 commented on src/main/java/homes/banzzokee/domain/shelter/entity/Shelter.java:
```diff
@@ -1,20 +1,28 @@
  * 보호소
  */
 @Entity
+@DynamicUpdate
```
처음 보는 어노테이션이라 아래 블로그를 살펴봤는데, 혹시 어떤 이유로 사용하셨는지 설명 가능하실까요?
https://multifrontgarden.tistory.com/299

eTaphee replied:
블로그에 나온 내용대로 변경된 내용에 대해서만 컬럼 업데이트를 하는게 좋다고 생각했었는데,

내용을 읽어보니 제가 정확히 이해하지못하고 `변경된 컬럼만 업데이트` 에 꽂혀서 무지성으로 사용했던거 같습니다.

정창님 덕분에 좀 더 찾아보고 공부해보겠습니다. 감사합니다.

eTaphee replied:
찾아보고 내용 공유해드릴게요.

eTaphee replied:
https://www.inflearn.com/questions/16697/%EC%97%85%EB%8D%B0%EC%9D%B4%ED%8A%B8-%EA%B3%A0%EA%B2%AC-%EA%B5%AC%ED%95%A9%EB%8B%88%EB%8B%A4

트레이드 오프를 생각하고 사용하면 될거같아요!

Shelter 테이블에 컬럼 또는 인덱스 컬럼이 많지 않아서 JPA 에서 생성한 `PrepareStatement` 를 그대로 사용하는 것이 나은거 같습니다.

`변경된 내용에 대해서만 컬럼을 업데이트 한다` 라는 말은 `@DynamicUpdate` 사용에 있어 합리적인 명분은 될 수 없겠네요.

nadoran78 replied:
상세한 답변 감사합니다!

nadoran78 reviewed (approved):
고생 많으셨습니다~

eTaphee committed:
Merge pull request #64 from banzzokee/feature/delete-shelter

feature: delete shelter

eTaphee committed:
Merge pull request #63 from banzzokee/feature/update-shelter

feature: update shelter
